Add assertions in CacheStorageCache to help debug threading issues

szewai opened the pull request:
#### 5c2607837be66195328f71725114deca9306f2dc
<pre>
Add assertions in CacheStorageCache to help debug threading issues
<a href="https://bugs.webkit.org/show_bug.cgi?id=253223">https://bugs.webkit.org/show_bug.cgi?id=253223</a>
rdar://106124999

Reviewed by Chris Dumez.

rdar://105122133 indicates CacheStorageCache may have threading issues (e.g. record url may become AtomString at some
point), so adding some assertion to help debug that.

This patch also make an isolated copy for record url before passing it to CacheStorageCache.

* Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
(WebKit::computeKeyURL):
(WebKit::CacheStorageCache::CacheStorageCache):
(WebKit::CacheStorageCache::getSize):
(WebKit::CacheStorageCache::open):
(WebKit::CacheStorageCache::retrieveRecords):
(WebKit::CacheStorageCache::removeRecords):
(WebKit::CacheStorageCache::findExistingRecord):
(WebKit::CacheStorageCache::putRecords):
(WebKit::CacheStorageCache::putRecordsAfterQuotaCheck):
(WebKit::CacheStorageCache::putRecordsInStore):
(WebKit::CacheStorageCache::removeAllRecords):
(WebKit::CacheStorageCache::close):
* Source/WebKit/NetworkProcess/storage/CacheStorageCache.h:
(WebKit::CacheStorageCache::assertIsOnCorrectQueue):

Canonical link: <a href="https://commits.webkit.org/261147@main">https://commits.webkit.org/261147@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07ae050028bb5428b153f6e50b8598e2fc98668a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110583 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19663 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43174 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1936 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119448 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21085 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10792 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/1416 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116325 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15705 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98896 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102900 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97665 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30563 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44021 "") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12315 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31901 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85934 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12875 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8852 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18239 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51517 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7721 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14759 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -38,9 +38,12 @@ namespace WebKit {
 
 static String computeKeyURL(const URL& url)
 {
+    RELEASE_ASSERT(url.isValid() && !url.isEmpty());
```
It is best to have separate assertions so we know which check is failing in case of a crash.

cdumez reviewed (commented):
Where is the isolatedCopy() speculative fix?

cdumez commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -375,8 +398,17 @@ void CacheStorageCache::removeAllRecords()
     m_isInitialized = false;
 }
+    
+void CacheStorageCache::assertIsOnCorrectQueue()
```
Could you mark this as inline? Just want to increase the likelihood we avoid function calls in release builds.

cdumez reviewed (approved):
LGTM with suggested changes.

szewai committed:
Add assertions in CacheStorageCache to help debug threading issues
https://bugs.webkit.org/show_bug.cgi?id=253223
rdar://106124999

Reviewed by Chris Dumez.

rdar://105122133 indicates CacheStorageCache may have threading issues (e.g. record url may become AtomString at some
point), so adding some assertion to help debug that.

This patch also make an isolated copy for record url before passing it to CacheStorageCache.

* Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
(WebKit::computeKeyURL):
(WebKit::CacheStorageCache::CacheStorageCache):
(WebKit::CacheStorageCache::getSize):
(WebKit::CacheStorageCache::open):
(WebKit::CacheStorageCache::retrieveRecords):
(WebKit::CacheStorageCache::removeRecords):
(WebKit::CacheStorageCache::findExistingRecord):
(WebKit::CacheStorageCache::putRecords):
(WebKit::CacheStorageCache::putRecordsAfterQuotaCheck):
(WebKit::CacheStorageCache::putRecordsInStore):
(WebKit::CacheStorageCache::removeAllRecords):
(WebKit::CacheStorageCache::close):
* Source/WebKit/NetworkProcess/storage/CacheStorageCache.h:
(WebKit::CacheStorageCache::assertIsOnCorrectQueue):

Canonical link: https://commits.webkit.org/261147@main
